Serialize Config operations

zslehoczky opened the pull request:
Collect all config operations from uapi and tun event loops into a channel
Create a single thread processing these operations (config worker)
Only instantiate the WireGuardConfig struct within this thread
Remove all concurrent types from WireGuardConfig
Use 1 thread / connection to read from & write to UAPI stream
Separate config operation parsing & execution
Add unit tests for parsing
Add new module: wg-uapi
Separate line reader from operation parsing, move line reader next to uapi worker
Move worker functions into separate files, make them testable by creating inner functions without thread spawning

zslehoczky committed:
Use channel instead of spawning new thread on each uapi event

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Replace map with loop

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Reorganize imports

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Serialize config operations in a single thread

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Only instantiate wireguard config inside config thread

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky commented:
I put this one back in draft because I realized I could extend this into fully serializing the config struct access from a single thread.

zslehoczky committed:
Remove Arc and Mutex from around Config

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

rot256 commented on wireguard-rs/src/run/runner.rs:
```diff
@@ -157,3 +168,30 @@ where
+            match message {
+                ConfigMessage::UapiStream(mut stream) => {
+                    uapi::handle(&mut stream, &mut wireguard_config);
+                }
```
Does this not mean that it hangs forever if someone makes a connection over the unix domain socket, but never closes it?

zslehoczky replied:
Hm, I'm not sure. To me it seems that handle() only ever processes a single instruction from the stream (256 characters in a single line). Is there an easy way to test this?

zslehoczky replied:
(I was under the impression that the UAPI interface is only used by configuration instructions coming from the same machine, which might be a totally wrong assumption.)

rot256 replied:
It is, but you could imagine a scenario where one application opened a steam, then kept it open, e.g. to obtain / track statistics. That would prevent another connection from configuring the interface.

Please correct me if I'm wrong, but I think we spawned a thread for every connection before this, which was also not great: as you said that means that accesses are not serialized (i.e. can be interleaved).

Maybe we should collect the whole request, then apply it and return the whole response.
It does mean that we can't stream the request/response....

rot256 replied:
Here is the Go implementation:

https://github.com/WireGuard/wireguard-go/blob/master/main.go#L245

Which is also not serialized (not ideal tbh).

rot256 replied:
Requests for the same connection are handled in a loop:

https://github.com/WireGuard/wireguard-go/blob/master/device/uapi.go#L428

rot256 replied:
But you are right: we are not crossing a trust boundary, so intentional DoS is not a grave concern. It's about whether multiple clients may be connected concurrently.

zslehoczky replied:
Yeah I see the problem now, so I think 1 thread/connection would be required, but then the messages themselves should go through the channel and be serialized. How does that sound?

rot256 replied:
Yea I that might be the way to go.

I haven't thought too hard about whether there is some way to make this nice with async without too much additional complexity.

zslehoczky replied:
Okay, so now we spawn 1 thread for each incoming connection, and we also keep the stream and thread alive for as long as the peer closes the UAPI connection (I believe previously we closed the stream and finished the thread after a single operation) so now the use case of multiple "listener" peers should work. (The peers should still send get messages every time they want to receive data, but now multiple parallel long-term streams are handled correctly.) I didn't test this though.

When a stream receives a valid operation, it sends it through the channel to be processed by the single worker thread. So this is the best of both worlds I think.

I also tried my best to add some sort of error recovery: each of the UAPI reader threads will read the stream for as long as they encounter an empty line. Then they will try to make sense of the lines: if they can, all good. If they can't they move on. This way we will never leave the stream in an inconsistent state, e.g. when an invalid key-value pair is encountered, we will still read until the empty line, so that we are able to call the next parse() at the beginning of the next message.

rot256 reviewed (commented):
Looks really nice.

I have some concerns about what happens if a client simply keeps the connection open. https://www.wireguard.com/xplatform/

zslehoczky committed:
Unnest helper functions from handle

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Move data into Set enum value

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add finalize fn

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Rename fn

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add BufReader

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add scoped thread for UAPI stream

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add TODOs

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add threaded parsing + serialized handling of UAPI messages

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix messages

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix format

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add lost fixes

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Handle tun up error

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Move thread creation into fn

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Save CI log

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix CI log saving

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Keep wg log after CI to save, delete at the end

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Don't delete log on linux also, remove cleanup step (automatically performed)

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add restore key for cache

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use a single bufreader object / thread

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

rot256 commented:
Is this good for me to re-review?

zslehoczky commented:
It’s supposed to be ready, however I’m having an issue on the ubuntu CI runner that I can’t reproduce locally, and which is probably a regression. I want to investigate further tomorrow. It has something to do with the Unix socket being closed from the beginning.

But in the meantime feel free to review it because this is the intended final form, based on what we discussed.

zslehoczky committed:
Use BufWriter, don't concatenate string

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fixes.

Use single string buffer for a stream.
Add error messages.
Parse trailing line for get operation.
Handle EOF on stream.

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Only save CI logs on GitHub if CI actually failed

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Update tuple syntax

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Move config operation into separate file

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Parsing fn now reads all lines until empty line

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

zslehoczky committed:
Use pattern matching instead of deref

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Remove Write from parser fn

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add unit tests to operation parsing

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky commented:
Turns out there were quite a few issues, but I believe this is now ready

zslehoczky committed:
Simplify line parsing loop

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

rot256 commented on .github/workflows/rust.yml:
```diff
@@ -33,6 +33,8 @@ jobs:
           ~/.cargo/bin
           target
         key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-cargo-
```
What's the significance of this?

zslehoczky replied:
In case there isn't an exact match for the cargo environment hash, then this allows the runner to load a previous cache (based on the prefix given) and so cargo can build the new environment iteratively, instead of from scratch

zslehoczky replied:
It's unrelated to the topic, only added it here because I had to iterate on debugging the CI on GitHub and that line is supposed to reduce build times

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

zslehoczky committed:
Fix clippy errors

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix clippy warnings

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Apply fmt

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky commented:
@rot256 Updated with the newest changes from develop, this is now ready for review again

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+    let lines: Vec<&str> = string_buffer
+        .lines()
+        .filter(|&line| !line.is_empty())
+        .collect();
```
There should be no empty lines?

zslehoczky replied:
We are parsing a single operation at a time, therefore there shouldn't be any empty lines (only the operation delimiter)

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+        } else {
+            return Ok(None); // EOF reached
+        }
+    }
```
I think a better option is to "parse" the entire request first into a set of operations:


```rust
enum Operation {
    FwMark(u32),
    PrivateKey(...),
    ...
}

enum Request {
   Get,
   Set(Vec<Operation>)
}

enum Response {
   Get(Vec<State>)
   Set(Status)
}
```

So that we can do all the request validation up front, instead of splitting it here and in the consumer.

I don't want to add it as a dependency, but an option is replicate something like: https://docs.rs/wireguard-uapi/3.0.1/wireguard_uapi/linux/set/struct.Peer.html

zslehoczky replied:
This is definitely better, but I think we could put this in a later PR

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+
+pub fn parse_config_operation<R: Read>(
+    reader: &mut BufReader<R>,
+    string_buffer: &mut String,
```
We can just create the buffer inside the function

zslehoczky replied:
This way we reuse a single buffer for the entire lifetime of a stream, leading to fewer allocations. But since this is not a performance-critical part of the program, we could also allocate a new one with every call and it probably wouldn't make a difference.
Or if we end up switching to the Lines approach, we wouldn't need the buffer at all, because then each line will be allocated separately.

zslehoczky replied:
I tried to switch to lines(), but turns out that it consumes self, and so it can't be invoked multiple times for the same stream. So because of that I stayed with the original solution using string buffer. (And we can't have a single Lines object for the entire loop, because then we can't write to the stream.)

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+        } else {
+            return Ok(None); // EOF reached
+        }
+    }
```
No need to collect to `string_buffer`, then parse, we can parse line-by-line:

```rust
fn parse_get<I: Iterator<Item = &'str>>(lines: I) -> Result<Vec<Operation>, Error> {
    let mut ops = vec![]
    for line in lines {
        if line.is_empty() {
           break
        }
        let (key, value) = line.split_once()?;
        match key {
            "private_key" => {
                ...
            }
        }
    }
    Ok(ops)
} 
```

zslehoczky replied:
This is much better, updated it

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+    let line_content_without_newline = &string_buffer[prev_len..(string_buffer.len() - 1)];
+
+    Ok(Some(line_content_without_newline))
+}
```
A reader should have `.lines()`?

https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html#a-more-efficient-approach

zslehoczky replied:
Right, but then it will allocate a new string for each line. But whichever is fine for me. Using lines() is much more concise, while the read_line approach reuses a single buffer so it's allocation-efficient.

zslehoczky replied:
lines() has its own problem: it consumes the reader, and because of that we can't use the stream for writing later. So because of that I kept the original reading

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+}
+
+fn parse_key_value_pair(ln: &str) -> Result<(String, String), ConfigError> {
+    let mut split = ln.splitn(2, '=');
```
`.split_once`: https://doc.rust-lang.org/std/primitive.str.html#method.split_once

zslehoczky replied:
Nice, much better

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+            _ => {
+                panic!();
+            }
+        }
```
`assert!(matches!` https://doc.rust-lang.org/std/macro.matches.html

zslehoczky replied:
Nice

rot256 commented on wireguard-rs/src/run/runner.rs:
```diff
@@ -155,3 +173,100 @@ where
+    stream: <<plt::UAPI as PlatformUAPI>::Bind as BindUAPI>::Stream,
+) -> ScopedJoinHandle<'scope, ()> {
+    thread_scope.spawn(|| {
+        let uapi_stream_sender = config_sender;
```
Why the alias?

zslehoczky replied:
It was needed because with scoped threads we can no longer use `move ||`, so variables needed to be moved explicitly

Anyway, now we use regular threads for this again, and so there is no need for explicit moving

rot256 commented on wireguard-rs/src/run/runner.rs:
```diff
@@ -155,3 +173,100 @@ where
+    config_sender: mpsc::Sender<ConfigMessage>,
+    stream: <<plt::UAPI as PlatformUAPI>::Bind as BindUAPI>::Stream,
+) -> ScopedJoinHandle<'scope, ()> {
+    thread_scope.spawn(|| {
```
I would spawn this in the connection handler, so this does not need threading, also allows independent testing: you don't need a thread scope.

zslehoczky replied:
Agreed, split it into 2 functions. Now there is a spawn* fn that also creates threading, and there is a worker fn which doesn't include threading which can be used in unit tests

rot256 commented on wg-uapi/src/uapi/set.rs:
```diff
@@ -52,55 +52,62 @@ impl<'a, C: Configuration> LineParser<'a, C> {
+        }
+    }
+
+    pub fn parse_line(&mut self, key: &str, value: &str) -> Result<(), ConfigError> {
+        #[cfg(debug_assertions)]
+        {
+            if !key.is_empty() {
+                log::debug!("UAPI: {}={}", key, value);
+            }
         }
```
We could separate the parsing from applying the effect of the configuration.

zslehoczky replied:
Definitely, that would be much nicer. This would also be consistent with your comments at the other file about parsing the operations semantically instead of just as structured strings

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,230 @@
+            _ => {
+                panic!();
+            }
+        }
```
If we implement `Eq` for a `Request` we can simplify this logic by `assert!`

zslehoczky replied:
Nice, thanks

rot256 commented on wireguard-rs/src/run/runner.rs:
```diff
@@ -142,37 +158,39 @@ fn spawn_tun_event_loop<'scope, 'env, T: Tun, B: PlatformUDP, S: Status>(
+                        tun_reader_jobs_running,
+                        config_sender.clone(),
+                        stream,
+                    );
```
Not sure I think this is simpler.

It makes it harder to test `uapi::handle` in isolation: the method needs to take `thread_scope` and `tun_reader_jobs_running`

zslehoczky replied:
Agreed, similarly to the above comment, I split it into 2 functions: one is used for spawning, the other can be used for unit tests.

(The spawning function is kept because it allows for nicer syntax than cloning + lambda creation)

rot256 reviewed (commented):
Idea: might make sense to separate the parsing response/request logic out of `wireguard-rs` and into `wg-uapi`: it could be useful to people outside of creating a WireGuard implementation, the same logic will be present in userspace/cli: https://crates.io/crates/wireguard-uapi

zslehoczky committed:
Remove thread scope and pass-by-ref from threads that are not using wireguard

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Move into non-scoped threads

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Refactor runner.rs

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Break up into testable functions

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Move spawning to inline

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use device reference in config

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Move UAPI to new module

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use split_once

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use assert matches

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix unit test Eq comparison

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add write response fn

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Update config operation parsing

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use result enum for more readable code

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use line iterator in message parsing

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Simplify line reader

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Add Eq to simplify unit test

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky commented:
@rot256  Thanks for the suggestions, now it's ready again for review. I have moved the code around a little bit: I created wg-uapi, and also I moved the worker functions from run.rs into their own files (it was a long overdue refactoring, run.rs was very long)

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

rot256 commented on wg-uapi/src/uapi/config_operation.rs:
```diff
@@ -0,0 +1,134 @@
+
+            for line in lines {
+                key_value_pairs.push(parse_key_value_pair(line)?);
+            }
```
```rust
lines.map(parse_key_value_pair).collect()
```

Rust knows how to collect:

```rust
Iter<Item = Result<T, E>>
```

Into:

```rust
Result<Vec<T>, E>
```

zslehoczky replied:
Done

rot256 commented on wireguard-rs/src/run/workers/line_reader.rs:
```diff
@@ -0,0 +1,137 @@
+            }
+            ReadLineResult::Eof => {
+                return ReadLinesResult::Eof;
+            }
```
Just a note, in case you did not know:

`loop` is special in Rust, it's different from `while true`...

The difference is that `loop` is divergent computation by default, which means: the `loop` itself, as an expression, can return *any* value (because it never terminates), e.g.

```rust
struct Cow {
   weight: u32,
   age: u32
}

fn example() -> Cow {
   loop {
       <anything>
   }
}
```

Works, no matter what `<anything>` is, as long as it does not have `break`:

In a `loop`, `break` is an expression which defines the value of the `loop` expression, e.g.

```rust
fn example() -> Cow {
   loop {
       break Cow {
           weight: 210,
           age: 10
       }
   }
}
```

So, e.g. 

```rust
pub fn read_line_block<R: Read>(
    reader: &mut BufReader<R>,
    string_buffer: &mut String,
) -> ReadLinesResult {
    loop {
        match read_line(reader, string_buffer) {
            ReadLineResult::Ok(line) => {
                if line.is_empty() {
                    break ReadLinesResult::Ok;
                }
            }
            ReadLineResult::Eof => {
                break ReadLinesResult::Eof;
            }
            ReadLineResult::Err(err) => {
                break ReadLinesResult::Err(err);
            }
        }
    }
}
```

Also works.

One might expect e.g. `for` to then return an `Option<ANY_TYPE>`, but for some reason, it does not...

zslehoczky replied:
Thanks, I've seen this somewhere in the code but didn't know how it worked. Very nice

rot256 commented on wireguard-rs/src/run/workers/line_reader.rs:
```diff
@@ -0,0 +1,137 @@
+    Ok,
+    Eof,
+    Err(Error),
+}
```
Are these errors or `Result`?

If it's an error, we should use `ReadLinesError` and return `Result<(), ReadLinesError>` to allow using e.g. the monadic notation (`?`).

zslehoczky replied:
There are 3 outcomes that I wanted to handle, that's why regular Result-like notation didn't seem like a good option (which represents a 2-way outcome).

There is Ok, which means the full block of lines could be received.
There is Eof, which means that the stream was closed before receiving a full block of lines. This is not an error, but a normal shutdown.
And there is Err, which means an unexpected issue with the stream, e.g an invalid UTF-8 character.

Do you think representing it as `Result<ReadLinesOutcome, io::Error>` would be better? Here ReadLinesOutcome would be an enum with the first 2 members (Ok, Eof)

zslehoczky replied:
Probably the alternative is better, as it allows normal error handling, as you mention.

zslehoczky replied:
Done

rot256 commented on wireguard-rs/src/run/workers/line_reader.rs:
```diff
@@ -0,0 +1,137 @@
+fn read_line<'buffer, R: Read>(
+    reader: &mut BufReader<R>,
+    string_buffer: &'buffer mut String,
+) -> ReadLineResult<'buffer> {
```
This should return a `Result`

zslehoczky replied:
Same as above, and quite painfully I couldn't even use the same enum. Although now that I think of it, I could have used a generic enum. Anyway, I will probably revise this and use a Result somehow.

zslehoczky replied:
Done

rot256 commented on wireguard-rs/src/run/workers/line_reader.rs:
```diff
@@ -0,0 +1,137 @@
+        }
+    };
+
+    ReadLineResult::Ok(line_content_without_newline)
```
Why not:

```rust
match string_buffer[prev_len..].strip_suffix('\n') {
    Some(line) => ReadLineResult::Ok(line)
    None => ReadLineResult::Eof,
};
```

zslehoczky replied:
Done

rot256 commented on wireguard-rs/src/run/workers/line_reader.rs:
```diff
@@ -0,0 +1,137 @@
+pub fn read_line_block<R: Read>(
+    reader: &mut BufReader<R>,
+    string_buffer: &mut String,
+) -> ReadLinesResult {
```
This function just reads all lines?

zslehoczky replied:
It reads a block of lines, which in this context is a sequence of non-empty lines followed by an empty line. This function is used to "atomically" read the minimal amount of processable information from the stream, i.e this function will block the thread until the full block is received, and only then will the thread proceed to actually parse the operation text.

This way we can handle stream shutdown (and in rare cases io errors) separately and up front. Any consequent processing can assume that a full block of text has been received without errors.

rot256 commented on wireguard-rs/src/run/workers/uapi.rs:
```diff
@@ -0,0 +1,160 @@
+    thread::spawn(move || {
+        uapi_server(uapi, config_sender);
+    })
+}
```
These functions are only used once? Does it make sense to inline them? The function signature is much bigger than the body.

zslehoczky replied:
Done, also for tun

rot256 commented on wireguard-rs/src/run/workers/uapi.rs:
```diff
@@ -0,0 +1,160 @@
+        let request_lines = request_buffer.lines().take_while(|&line| !line.is_empty());
+
+        let response = parse_config_operation(request_lines).and_then(|config_operation| {
+            let (config_result_sender, config_result_receiver) = crossbeam_channel::unbounded();
```
It could be capacity 1

zslehoczky replied:
Done

rot256 commented on wireguard-rs/src/run/workers/uapi.rs:
```diff
@@ -0,0 +1,160 @@
+                .recv()
+                .expect("channel should be open until result is received")
+        });
+
```
Consider refactoring this into a `handle_request` function: that way, returning early with an error becomes very easy, ala.

```rust
fn handle_request(...) -> Result<(), Error> {
   let req = read_line_block(&mut reader)?;
   let request_lines = request_buffer.lines().take_while(|&line| !line.is_empty());
   let config_operation = parse_config_operation(request_lines)?;
   let (config_result_sender, config_result_receiver) = crossbeam_channel::unbounded();
   ...
}
```

Then calling it in a loop.

zslehoczky replied:
Done, except the first error needs conversion into ConfigError, so ? can't be used there

rot256 reviewed (dismissed):
Nice!

My primary question would be: can we get rid of the mutable buffers which makes things harder to read / reason about and just return new `Strings/Vec<u8>` with the small performance overhead it entails.

This code is not super performance critical, as long as it has the right asymptotics e.g. no accidental O(n^2) behavior.

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into uapi-mpsc

zslehoczky committed:
Add bounded channel with capacity of 1

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Inline one liner spawn functions

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use functional idiom

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Use result in line reader

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Refactor uapi handler for better error handling

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>
